utils/s3/client.cc

```diff
@@ -592,6 +592,99 @@ class client::multipart_upload {
     unsigned parts_count() const noexcept { return _part_etags.size(); }
 };
 
+class client::copy_s3_object final : public multipart_upload {
+public:
+    copy_s3_object(shared_ptr<client> cln, sstring source_object, sstring target_object, std::optional<tag> tag, abort_source* as)
+        : multipart_upload(std::move(cln), std::move(target_object), std::move(tag), as), _source_object(std::move(source_object)) {}
+
+    future<> copy() {
+        auto source_size = co_await _client->get_object_size(_source_object);
+        if (source_size <= _max_copy_part_size) {
+            co_await copy_put();
+        } else {
+            co_await copy_multipart(source_size);
+        }
+    }
+
+private:
+    future<> copy_put() {
+        auto req = http::request::make("PUT", _client->_host, _object_name);
+        if (_tag) {
+            req._headers["x-amz-tagging"] = seastar::format("{}={}", _tag->key, _tag->value);
+        }
+        req._headers["x-amz-copy-source"] = _source_object;
+
+        co_await _client->make_request(std::move(req), ignore_reply, http::reply::status_type::ok, _as);
+    }
+
+    future<> copy_multipart(size_t source_size) {
+        co_await start_upload();
+        auto part_size = _max_copy_part_size;
+        std::exception_ptr ex;
+
+        try {
+            for (size_t offset = 0; offset < source_size; offset += part_size) {
+                part_size = std::min(source_size - offset, part_size);
+                co_await copy_part(offset, part_size);
+            }
+
+            co_await finalize_upload();
+        } catch (...) {
+            ex = std::current_exception();
+        }
+        if (ex) {
+            if (!_bg_flushes.is_closed()) {
+                co_await _bg_flushes.close();
+            }
+            co_await abort_upload();
+            std::rethrow_exception(ex);
```

Comment:
If we failed, do we leave uploaded parts? And what if we crash, how are these garbage-collected?

Reply:
by object lifecycle management policy on the bucket, this case is not different from any other multipart upload we are already running

Reply:
Ok, then consider the question expanded to other multipart uploads too. What happens to abandoned uploads?

Reply:
The answer is still the same, by object lifecycle management policy on the bucket. I think we already discussed it, or maybe it was @mykaul ? in any case AFAIR nowadays the ScyllaManager uses `rclone` to cleanup the dangling multipart uploads and want to reiterate it once again, we MUST set the object lifecycle management to take care if it.

Reply:
ok.